pkg/controller/atlasdeployment/atlasdeployment_controller.go

```diff
@@ -172,26 +160,88 @@ func (r *AtlasDeploymentReconciler) Reconcile(context context.Context, req ctrl.
 
 	switch {
 	case existsInAtlas && wasDeleted:
-		return r.delete(ctx, deploymentInAKO)
+		return r.delete(workflowCtx, deploymentInAKO)
 	case !existsInAtlas && wasDeleted:
-		return r.unmanage(ctx, atlasDeployment)
+		return r.unmanage(workflowCtx, atlasDeployment)
 	case !wasDeleted && isServerless:
 		var serverlessDeployment *deployment.Serverless
 		if existsInAtlas {
 			serverlessDeployment = deploymentInAtlas.(*deployment.Serverless)
 		}
-		return r.handleServerlessInstance(ctx, deploymentInAKO.(*deployment.Serverless), serverlessDeployment)
+		return r.handleServerlessInstance(workflowCtx, deploymentInAKO.(*deployment.Serverless), serverlessDeployment)
 	case !wasDeleted && !isServerless:
 		var clusterDeployment *deployment.Cluster
 		if existsInAtlas {
 			clusterDeployment = deploymentInAtlas.(*deployment.Cluster)
 		}
-		return r.handleAdvancedDeployment(ctx, deploymentInAKO.(*deployment.Cluster), clusterDeployment)
+		return r.handleAdvancedDeployment(workflowCtx, deploymentInAKO.(*deployment.Cluster), clusterDeployment)
 	}
 
 	return workflow.OK().ReconcileResult(), nil
 }
 
+func (r *AtlasDeploymentReconciler) getProjectFromAtlas(ctx *workflow.Context, atlasDeployment *akov2.AtlasDeployment) (*project.Project, error) {
```

Comment:
is this the very same method as from the database users PR? If so, does it make sense to somehow centralize this method?

Reply:
The internal services built and set are distinct from controller to controller. Also, deployment specifically still needs the old Atlas client which needs to be initialized. I think duplication here is better than the complex function that will try to cover every case on every controller.

Reply:
Wait even better, setup the client(s) and services first of all in a method:
```go
func (r *AtlasDeploymentReconciler) setupDeploymentClientServices(ctx *workflow.Context, credentialsName, namespace string) error
```

After this the reconciler only needs to call `r.projectService.GetProject`

On the other codepath you cal the `setupDeploymentClientServices` first jus the same, then call a functions to resolve the project from Kubernetes. A function that can be shared across controllers.

The only thing that needs to be per reconciler is the clients and services setup from credentials.